### ISSUE TITLE
Document dangers of using knex.raw with query.toString

### ISSUE THREAD

**Issue by ben-turner** [knex/documentation#227]
I have been using Knex for a couple of years now and only just noticed that using the output of toString in a new raw query is a SQL injection vulnerability. Maybe I didn't RTFM well enough, but it seems like many other people are doing the same thing:

https://github.com/tgriesser/knex/issues/701
https://jaketrent.com/post/upsert-knexjs/

`toString()` is a pretty standard function in JavaScript—I came across it without seeing it in the docs—and it seems logical that it would output safe SQL if you don't see the warning in the docs.

Adding a warning about this pattern in the `raw` docs might make it more visible, or adding a section about common mistakes and gotcha's could also help.

An example of when one might do this:
```
const foo = 'bar?';
const query = knex('sometable').insert({ foo });
knex.raw(`${query.toString()} ON CONFLICT DO NOTHING`);
```

**Comment by elhigu** [comment]
It is documented yeah and still I need to tell people again and again in stackoverflow and github issues that they should never stringify query unless it is for debugging. 

Solution to this would be to deprecate `.toString()` and `.toQuery()` methods all together and rename them to something like `.queryAsStringUnsafeForDebuggingOnly()`.

I don't believe more documentation will help, since people doesn't read it.

**Comment by elhigu** [comment]
Looks like the warnings could be written with bigger red flags though http://knexjs.org/#Interfaces-toString

**Comment by felixmosh** [comment]
is a `query.toString()` of a builded query returns a vulnerable query?

**Comment by ben-turner** [comment]
@felixmosh Yes. `toString()` interpolates variables directly into the query string, rather than sending them as parameters.

**Comment by felixmosh** [comment]
I can't produce a SQLi with it.
```javascript
const sql = knex
  .insert({ id: 2, name: "t`')est" })
  .into('persons')
  .toString();
console.log(sql); // insert into `persons` (`id`, `name`) values (2, 't`\\')est')
```
This is a valid query, with a proper escaping.

**Comment by elhigu** [comment]
@felixmosh toString() doesn't work the same way as query in some other cases as well when it tries to stringify objects etc.. Main reason why knex says it is insecure is because knex cannot guarantee that escaping works correctly in every case (there are some measures trying to escape strings though). I'm pretty sure there are some smart hackerpeople who can break it.

**Comment by felixmosh** [comment]
What is a better approach to introduce un supported syntax?
I'm the author of [knex-on-duplicate-update](https://github.com/felixmosh/knex-on-duplicate-update), this is why I have concerns :]

**Comment by ben-turner** [comment]
My suggestion is that these functions be removed entirely. I brought this up in the main Knex repo [here](https://github.com/knex/knex/issues/3377)

**Comment by ben-turner** [comment]
@felixmosh I submitted a PR to your repo. A more secure workaround I've found is to use `toSQL` like so.

```
const {
  sql,
  bindings,
} = originalQuery.toSQL();

const newQuery = knex.raw(
  `${sql} on duplicate key update key=?`,
  [ ...bindings, newValue ]
);
```
